### DISCUSSION TITLE
How do you model exclusive relations on a model?

### DISCUSSION THREAD

**Discussion by binajmen** [prisma/prisma#13196] [Q&A] [answered]
This is not entirely prisma related, but I'm asking the question anyway:

Is it wrong to model the `Interval` as follow?

```prisma
model Location {
  id   String @id @default(cuid())
  name String

  intervals Interval[]
}

model Category {
  id   String @id @default(cuid())
  name String

  events Event[]
}

model Interval {
  id   Int    @id @default(autoincrement())
  day  Int    @db.SmallInt // dow – Sunday (0) to Saturday (6)
  from String @db.Char(5) // '07:00'
  to   String @db.Char(5) // '17:00'

  Location   Location? @relation(fields: [locationId], references: [id])
  locationId String?
  Event      Event?    @relation(fields: [eventId], references: [id])
  eventId    Int?
}

model Event {
  id          Int        @id @default(autoincrement())
  name        String
  type        EventType
  intervals   Interval[]

  Category   Category? @relation(fields: [categoryId], references: [id])
  categoryId String?
}
```

What feels wrong is the exclusive relations: `locationId` or `eventId`, never both + the fact that they are both optionals, whereas in reality I should always have one relation set.

What if `Interval` has more relations (it is in my current project)? I would end up with a table schema with a lot of `null` values. Is it acceptable? Is there a better way to model that?

I'd love to hear your feedbacks :)

<img width="700" alt="Screenshot 2022-05-05 at 16 32 55" src="https://user-images.githubusercontent.com/15611419/166946733-39e051de-e686-4a6c-b054-35f90ce1971d.png">

**Comment by michaeldebetaz** [comment]
I'm facing the same problem. Fundamentally, I want to link a `Password` model to two other models, for example `Student` and `Teacher`.

I don't know if it exists  way to extend a model (like having a base `User`, which has a password and other common user data, and having it linked to 2 different models, `Student` and `Teacher`, to store their specific data). 

```prisma
model User {
  id         String    @id @default(cuid())
  email      String    @unique
  password   Password?
  passwordId String?
}
 
model Password {
  hash   String
  user   User   @relation(fields: [userId], references: [id])
  userId String @unique
}

model Student { ... }

model Teacher { ... }

```
I don't think it's a new problem and I would love to hear some tips 😀 Also I don't think  it should be a good idea to get queries like `db.student.studentPassword...`. I see potential solutions, but they look wrong somehow.

**Comment by binajmen** [comment, marked as answer]
After a fresh night, I'm wondering if implicit m-to-n relationships could be the solution:
```prisma
model Interval {
  id   Int    @id @default(autoincrement())
  day  Int    @db.SmallInt // dow – Sunday (0) to Saturday (6)
  from String @db.Char(5) // '07:00'
  to   String @db.Char(5) // '17:00'

  locations Location[]
  events    Event[]
}
```
This will create bridge tables, and the table `interval` will not be polluted. I would then query things like:
```ts
  const date = new Date();
  const dayOfWeek = date.getDay();
  const time = date.getHours() + ":" + date.getMinutes();

  // find locations that have no time restrictions or have an active interval
  const locations = await prisma.location.findMany({
    where: {
      OR: [
        { intervals: { none: {} } },
        { intervals: { some: { day: dayOfWeek, from: { lte: time }, to: { gte: time } } } },
      ],
    },
  });
```
It works if you don't care about the cardinality.

**Comment by michaeldebetaz** [comment]
After playing aroung in my case, I think that logic duplication is easier to manage. That means that I will have a `StaffMember` and an `Student` model, and have respectively a 1-to-1 `StaffMemberPassword` and `StudentPassword`. I was afraid of creating ugly queries, but I can keep the same column names accross the models. So, I can still have `db.tutor.password` and so.

Here are my current models (simplified, for the sake of the explanation):

```prisma
model StaffMember {
  id        String               @id @default(cuid())
  email     String               @unique
  password  StaffMemberPassword?
  students Student[]
}

model StaffMemberPassword {
  hash          String
  staffMember   StaffMember @relation(fields: [staffMemberId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  staffMemberId String      @unique
}

model Student {
  id        String           @id @default(cuid())
  email     String           @unique
  password  StudentPassword?
  manager   StaffMember      @relation(fields: [managerId], references: [id], onDelete: Restrict, onUpdate: Cascade)
  managerId String
}

model StudentPassword {
  hash      String
  student   Student @relation(fields: [studentId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  studentId String  @unique
}

```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
